Tools: Added option of --why/Y to crm_resource.

aravind-kumar opened the pull request:
WIP and Yet to test.

Wanted to know if this is the correct approach.

1.Added a new option "Why" with an alias of 'Y'
2.Checking if rsc_cmd == 'Y'
    2.1 Checking if rsc id is present
    2.2 If so find the resource and do a cli_resource_check

Should I change the cli_resource_check for all the resources if no rsc id mentioned.? 
If so should I be doing it by iterating data_set.resources ?

kgaillot commented:
Yes, this is what I had in mind. Some comments:

* Use four spaces (no tabs) as the indent. Most editors have a setting to make that automatic.
* I do think it would be nice to check all resources if no rsc_id is given. You can see an example of iterating over data_set.resources in the 'l' and 'A' handling.
* cli_resource_check() prints a message only if there's a reason the resource should be stopped. I think we should also print "Resource is active" or "Resource is stopped" first. See the resource_is_running_on() function for an example usage of rsc->fns->location() to check what hosts a given resource is active on (the list will be empty if the resource is stopped). It might be interesting to check the -N option to limit the active/stopped check to that one node, if the user specified it (but still check on all nodes if -N was not specified).

aravind-kumar commented:
Thanks, Ken. 😄 

- Made changes to iterate through data_set.resources if no resources are mentioned.
- Made changes to print "Resource is active" or "Resource is stopped" and only if a resource is seen to be stopped, proceed with cli_resource_check()
- Added Code to limit the active/stopped check to the node specified by -N option.
- Is it possible to specify a list of resources using the -N option?
-With Both Resource_id option and -N option ...there were 4 possible sets of outcomes.
Have handled all of them.

Please Let me know if its ok ? 😄 

Also am yet to test the code.
Could you tell me possible ways to test ?
Should i write a dummy RA which fails at all times and try to check the reason with --why..?

kgaillot commented on tools/crm_resource.c:
```diff
@@ -802,50 +802,7 @@ main(int argc, char **argv)
-            g_list_free(hosts);
-            hosts = NULL;
-	 } 
+      cli_resource_why(cib_conn,data_set.resources,rsc_id,host_uname);
```
This looks good.

Be sure to go through your changes and make sure all the indenting is correct: no tabs, and four spaces per indent.

kgaillot commented on tools/crm_resource_runtime.c:
```diff
@@ -1750,3 +1752,87 @@ cli_resource_move(const char *rsc_id, const char *host_name, cib_t * cib, pe_wor
+            cli_resource_check(cib_conn, rsc);
+	} else {
+	    printf("Resource %s is running",rsc->id);
+	}
```
cli_resource_check() will actually be useful whether or not the resource is running, so we can move it below this if-else block.

kgaillot commented on tools/crm_resource_runtime.c:
```diff
@@ -1750,3 +1752,87 @@ cli_resource_move(const char *rsc_id, const char *host_name, cib_t * cib, pe_wor
+        cli_resource_check(cib_conn, rsc);
+    } else {
+        printf("Resource %s is not running on host %s",rsc->id,host_uname);
+    }
```
Similarly here, let's print a message in both cases, and run cli_resource_check() after the if-else block.

kgaillot commented on tools/crm_resource_runtime.c:
```diff
@@ -1750,3 +1752,87 @@ cli_resource_move(const char *rsc_id, const char *host_name, cib_t * cib, pe_wor
+    for(lpc = unactiveResources; lpc != NULL; lpc = lpc->next) {
+        resource_t *rsc = (resource_t *) lpc->data;
+        cli_resource_check(cib_conn,rsc);
+     }
```
After thinking more about this, I think we should do two loops, through active and inactive resources. The active loop can print the "is running on" message, and the inactive loop can print that the resource "is assigned to NODE-NAME but not running". Both loops should call cli_resource_check().

That way, if someone runs "crm_resource --why -N node1", they get a full list of all resources assigned to the node, and whether each one is running or not running (and possibly why or why not).

aravind-kumar replied:
Yeah this seems to be a nice idea,
Would change it 😄

kgaillot commented on tools/crm_resource_runtime.c:
```diff
@@ -1750,3 +1752,87 @@ cli_resource_move(const char *rsc_id, const char *host_name, cib_t * cib, pe_wor
+        cli_resource_check(cib_conn, rsc);
+    } else {
+	printf("Resource %s is running",rsc->id);
+    }
```
Similarly here, cli_resource_check() can be moved after the if-else block.

aravind-kumar replied:
sry my bad :slightly_smiling_face:

kgaillot commented on tools/crm_resource_runtime.c:
```diff
@@ -1713,3 +1752,87 @@ cli_resource_move(const char *rsc_id, const char *host_name, cib_t * cib, pe_wor
+static void 
+cli_resource_why_without_rsc_with_host(cib_t *cib_conn,GListPtr resources,const char* host_uname)
+{
+    GListPtr allResources = get_all_resources(host_uname,resources);
```
I think we can avoid creating a new function to get all resources. You can find the node in the data set using pe_find_node() (see the 'U' handling for an example), and pass the node entry rather than the host name to cli_resource_why(). There is a node->details->allocated_rsc member that is a list of all resources assigned to that node (whether or not they are running). There is also node->details->running_rsc with just the ones that are running, so you can use subtract_lists() to get the inactive ones. The node name is available as node->details->uname.

aravind-kumar replied:
Nice ! I did not know that we could get it from the node_t. :slightly_smiling_face: 
This new apporach looks much cleaner and also removes the redudant code.

kgaillot commented on tools/crm_resource_runtime.c:
```diff
@@ -1713,3 +1752,87 @@ cli_resource_move(const char *rsc_id, const char *host_name, cib_t * cib, pe_wor
+     }
+
+     g_list_free(hosts);
+     hosts = NULL;
```
This free+null needs to be inside the for loop, otherwise the result from the previous iteration will affect the current one.

aravind-kumar replied:
Nice Catch , thank you :slightly_smiling_face: . I had that in mind but i coded it wrong.Thank you :slightly_smiling_face:

aravind-kumar commented:
Thanks Ken 😄 
They were great review comments :slightly_smiling_face: 
Following sets of changes has been made

- Refactored the code by removing the redudant function get_all_resources() and rather using the node_t object to find the list of all and active resources.

- Fixed the leak by freeing the list of hosts while iterating through the resouce list when both resource and host name aren't provided. (thanks for that 👍 )

- Made cli_resource_check() to be executed in regardless of if a resource is running or not.Had earlier missed a few places.

- Changed the prints to print the resource name and its assigned hostname.

-

kgaillot commented:
Looks great! Let me know how testing went. One last request: Please squash the commits into one, and use a commit message following our style. If you're not familiar with squashing commits, see:

  http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

For the commit message, start the summary line with "Feature: tools:" and then describe the change. When it comes time to make a new Pacemaker release, that "Feature:" keyword will flag this as something to go into the change log.

After squashing, you can push back to this pull request with:

    git push origin +HEAD:stonithMaxAttempts

which will remove the old commits from here and push the new one.

aravind-kumar commented:
Hey Ken, 
I get the following error while i was trying to test
crm_resource --why
Argument code 0131 (Y) is not (?yet?) supported
Invalid option(s) supplied, use --help for valid usage


Thoughts ?

kgaillot commented:
> Hey Ken,
> I get the following error while i was trying to test
> crm_resource --why
> Argument code 0131 (Y) is not (?yet?) supported
> Invalid option(s) supplied, use --help for valid usage
> 
> Thoughts ?

Oh right, I missed that you aren't actually parsing the option.

See the while loop beginning at line 299. This is where each option is recognized and remembers any arguments it needs. Since 'Y' doesn't need any arguments, it can go with other such options around 'L'.

aravind-kumar commented:
Hi Ken

Apologies for the delay.

I was testing the code and it was working for the cases wherein the resource is running properly with and without node list being given.

However, I also wanted to test out a failure case.
1.I had tried to use target-role and is-managed and they seem to work with appropriate error messages such as 
 * The configuration specifies that 'Dummy2' should remain stopped -> Target-role=stopped
 * The configuration prevents the cluster from stopping or starting 'Dummy3' (unmanaged) -> For Unmanaged Resource.


However, i also observed the following
1.I had created a resource Dummy1
2.Deleted the resource using pcs resource delete Dummy1
3.Created Dummy2 and Dummy3 resources
4.Modified the target-role for Dummy2 and Changed Dummy3 to Unmanaged
5.Ran crm_resource --why

The output however also displays that Dummy1 isn't running.
Following is the output.
-----------------------------------------------------------------------------------
[root@TPC-BB5-28-020 tools]# pcs status
Cluster name: HA
WARNING: no stonith devices and stonith-enabled is not false
Stack: corosync
Current DC: TPC-BB5-28-015.lab.blr.sandvine.com (version 1.1.15-11.el7-e174ec8) - partition with quorum
Last updated: Mon May 29 19:46:47 2017          Last change: Mon May 29 19:46:40 2017 by root via cibadmin on TPC-BB5-28-020.lab.blr.sandvin$.com

2 nodes and 2 resources configured: 1 resource DISABLED and 0 BLOCKED from being started due to failures

Online: [ TPC-BB5-28-015.lab.blr.sandvine.com TPC-BB5-28-020.lab.blr.sandvine.com ]

Full list of resources:

 Dummy2 (ocf::heartbeat:Dummy): Stopped (disabled)
 Dummy3 (ocf::heartbeat:Dummy): Stopped (unmanaged)

Daemon Status:
  corosync: active/disabled
  pacemaker: active/disabled
  pcsd: active/disabled
---------------------------------------------------------------------------------------------
As shown above Dummy1 is not a resource.However, crm_resource is displaying Dummy1 as well.
----------------------------------------------------------------------------------------------
[root@TPC-BB5-28-020 tools]# crm_resource --why

Resource Dummy2 is not running
 Resource role is Stopped (Please Ignore this statement added for debugging)
  * The configuration specifies that 'Dummy2' should remain stopped


Resource Dummy3 is not running
 Resource role is (null) (Please Ignore this statement added for debugging)
  * The configuration prevents the cluster from stopping or starting 'Dummy3' (unmanaged)


Resource Dummy1 is not running
 Resource role is (null) (Please Ignore this statement added for debugging)
---------------------------------------------------------------------------------------------------

Could you please tell me if its expected?
I belive data_set.resources crm_resource.c:815 inside handling of 'Y' also contains the deleted resource?

kgaillot commented:
That would mean Dummy1 is still referenced in the CIB somewhere. Try "cibadmin -Q | less" and search for Dummy1 to see where it is. I'd expect "pcs resource delete Dummy1" to erase it entirely, so I'm not sure why it would still be in there. In any case, it's OK for "crm_resource --why" to show it if it's there.

The newlines should go at the end of the printf() strings rather than the beginning. Otherwise the last line printed won't have one.

aravind-kumar commented:
Thanks Ken :)
Changed the newlines to the end of the file.
Also squashed all the commits to a single commit message.
